### PR TITLE
Created TopicQos class [5982]

### PR DESCRIPTION
**Pull request by AlvaroGonzalezMolinillo**

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/0/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/0/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/0/)

File: include/fastdds/topic/qos/TopicQos.hpp [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+namespace eprosima {
+namespace fastdds {
+
+/**
+ * Class TopicQos, containing all the possible Qos that can be set for a determined Topic.
+ * Although these values can be set and are transmitted
+ * during the Endpoint Discovery Protocol, not all of the behaviour associated with them has been implemented in the library.
+ * Please consult each of them to check for implementation details and default values.
+ * @ingroup FASTRTPS_ATTRIBUTES_MODULE
```
Comment (raquelalvarezbanos) [inline include/fastdds/topic/qos/TopicQos.hpp]: Maybe `FASTDDS_QOS_MODULE` would be a better name? Double-check with @LuisGP

File: include/fastdds/topic/qos/TopicQos.hpp [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+               (this->m_destinationOrder == b.m_destinationOrder) &&        
+               (this->m_history == b.m_history) &&        
+               (this->m_resourceLimits == b.m_resourceLimits) &&        
+               (this->m_transportPriority == b.m_transportPriority) &&        
+               (this->m_lifespan == b.m_lifespan) &&        
+               (this->m_ownership == b.m_ownership);
+    }
+
+    //!Topic Data Qos, NOT implemented in the library.
+    fastrtps::TopicDataQosPolicy m_topicData;
```
Comment (raquelalvarezbanos) [inline include/fastdds/topic/qos/TopicQos.hpp]: I know you probably copied this from an existing Qos struct, but from now on, please follow the coding style: https://github.com/richiware/cppstyle/blob/master/README.es.md both for member variables and methods

File: include/fastdds/topic/qos/TopicQos.hpp [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    //!History Qos
+    fastrtps::HistoryQosPolicy m_history;
+
+    //!Resource Limits Qos
+    fastrtps::ResourceLimitsQosPolicy m_resourceLimits;
+
+    //!Transport Priority Qos
+    fastrtps::TransportPriorityQosPolicy m_transportPriority;
+
+    //!Lifespan Qos, NOT implemented in the library.
```
Comment (raquelalvarezbanos) [inline include/fastdds/topic/qos/TopicQos.hpp]: `Lifespan` is in fact implemented

File: include/fastdds/topic/qos/TopicQos.hpp [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+     */
+    RTPS_DllAPI void setQos(const TopicQos& qos, bool first_time);
+
+    /**
+     * Check if the Qos values are compatible between each other.
+     * @return True if correct.
+     */
+    RTPS_DllAPI bool checkQos() const;
+
+    RTPS_DllAPI bool canQosBeUpdated(const TopicQos& qos) const;
```
Comment (raquelalvarezbanos) [inline include/fastdds/topic/qos/TopicQos.hpp]: Please document this method

File: src/cpp/fastdds/topic/qos/TopicQos.cpp [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+{
+
+}
+
+TopicQos::~TopicQos()
+{
+
+}
+
+void TopicQos::setQos(const TopicQos& qos, bool first_time)
```
Comment (raquelalvarezbanos) [inline src/cpp/fastdds/topic/qos/TopicQos.cpp]: Remove these methods if they are not implemented. You can leave a `TODO` comment in the header file

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/133/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/122/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/132/)

File: include/fastdds/dds/topic/qos/TopicQos.hpp [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+    /* TODO: Implement this method
+     * Check if the Qos can be update with the values provided. This method DOES NOT update anything.
+     * @param qos Reference to the new qos.
+     * @return True if they can be updated.
+    RTPS_DllAPI bool canQosBeUpdated(const TopicQos& qos) const;
+    */
+};
+
+extern TopicQos TOPIC_QOS_DEFAULT;
```
Comment (LuisGP) [inline include/fastdds/dds/topic/qos/TopicQos.hpp]:
```suggestion
extern const TopicQos TOPIC_QOS_DEFAULT;
```

File: src/cpp/dds/topic/qos/TopicQos.cpp [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+/**
+ * @file TopicQos.cpp
+ *
+ */
+
+#include <fastdds/dds/topic/qos/TopicQos.hpp>
+
+using namespace eprosima::fastdds;
+
+TopicQos eprosima::fastdds::TOPIC_QOS_DEFAULT;
```
Comment (LuisGP) [inline src/cpp/dds/topic/qos/TopicQos.cpp]:
```suggestion
const TopicQos eprosima::fastdds::TOPIC_QOS_DEFAULT;
```

File: include/fastdds/dds/topic/qos/TopicQos.hpp [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+ */
+
+
+#ifndef _FASTDDS_TOPICQOS_HPP
+#define _FASTDDS_TOPICQOS_HPP
+
+#include <fastrtps/qos/QosPolicies.h>
+
+namespace eprosima {
+namespace fastdds {
```
Comment (LuisGP) [inline include/fastdds/dds/topic/qos/TopicQos.hpp]: Add `namespace dds`.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/257/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/234/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/256/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/347/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/320/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/345/)

File: include/fastdds/dds/topic/qos/TopicQos.hpp (hunk 8 above)
Comment (laura-eprosima) [reply include/fastdds/dds/topic/qos/TopicQos.hpp]: Added

File: include/fastdds/dds/topic/qos/TopicQos.hpp (hunk 6 above)
Comment (laura-eprosima) [reply include/fastdds/dds/topic/qos/TopicQos.hpp]: Changed

File: src/cpp/dds/topic/qos/TopicQos.cpp (hunk 7 above)
Comment (laura-eprosima) [reply src/cpp/dds/topic/qos/TopicQos.cpp]: Changed

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/353/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/326/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/351/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/368/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/336/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/366/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/497/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/449/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/494/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/525/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/468/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/521/)

File: include/fastrtps/qos/QosPolicies.h [hunk 9]
```diff
@@ -784,6 +784,16 @@ class ResourceLimitsQosPolicy : public Parameter_t, public QosPolicy
 
     virtual RTPS_DllAPI ~ResourceLimitsQosPolicy() { }
 
+    bool operator==(const ResourceLimitsQosPolicy& b) const
```
Comment (LuisGP) [inline include/fastrtps/qos/QosPolicies.h]:
```suggestion
    bool operator ==(
            const ResourceLimitsQosPolicy& b) const
```

File: include/fastrtps/qos/QosPolicies.h [hunk 10]
```diff
@@ -930,6 +940,14 @@ class TransportPriorityQosPolicy : public Parameter_t , public QosPolicy
         uint32_t value;
         RTPS_DllAPI TransportPriorityQosPolicy():Parameter_t(PID_TRANSPORT_PRIORITY,4),QosPolicy(false),value(0){};
         virtual RTPS_DllAPI ~TransportPriorityQosPolicy(){};
+
+        bool operator==(const TransportPriorityQosPolicy& b) const
```
Comment (LuisGP) [inline include/fastrtps/qos/QosPolicies.h]:
```suggestion
        bool operator ==(
                const TransportPriorityQosPolicy& b) const
```

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/581/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/507/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/577/)

**Comment by LuisGP** [comment]
> * Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/507/)

Mac failed dut to the IO error. Please @raquelalvarezbanos, merge this PR.

**Comment by raquelalvarezbanos** [comment]
Merging, as Mac failure is unrelated and passed in the previous build.

**Pull request merged**